rollup/cmd/permissionless_batches/app/app.go

```diff
@@ -0,0 +1,335 @@
+	latestL2Block, err := l2Watcher.Client.BlockNumber(context.Background())
+	if err != nil {
+		return 0, fmt.Errorf("failed to get latest L2 block number: %w", err)
+	}
+
+	log.Info("Latest L2 block number", "latest L2 block", latestL2Block)
+
+	if l2BlockHeightLimit > latestL2Block {
+		return 0, fmt.Errorf("l2BlockHeightLimit is higher than the latest L2 block number, not all blocks are available in L2geth: %d > %d", l2BlockHeightLimit, latestL2Block)
+	}
+
+	toBlock := latestL2Block
+	if l2BlockHeightLimit > 0 {
+		toBlock = l2BlockHeightLimit
+	}
+
+	err = l2Watcher.GetAndStoreBlocks(context.Background(), fromBlock, toBlock)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get and store blocks: %w", err)
+	}
+
+	log.Info("Fetched L2 blocks from", "fromBlock", fromBlock, "toBlock", toBlock)
+
+	return toBlock, nil
+}
+
+// restoreMinimalPreviousState restores the minimal previous state required to be able to create new chunks, batches and bundles.
+func restoreMinimalPreviousState(cfg *config.Config, chunkProposer *watcher.ChunkProposer, batchProposer *watcher.BatchProposer) (*orm.Chunk, *orm.Batch, error) {
+	log.Info("Restoring previous state with", "L1 block height", cfg.RecoveryConfig.L1BlockHeight, "latest finalized batch", cfg.RecoveryConfig.LatestFinalizedBatch)
+
+	// TODO: make these parameters -> part of genesis config?
+	scrollChainAddress := common.HexToAddress("0x2D567EcE699Eabe5afCd141eDB7A4f2D0D6ce8a0")
+	l1MessageQueueAddress := common.HexToAddress("0xF0B2293F5D834eAe920c6974D50957A1732de763")
+
+	l1Client, err := ethclient.Dial(cfg.L1Config.Endpoint)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to connect to L1 client: %w", err)
+	}
+	reader, err := l1.NewReader(context.Background(), l1.Config{
+		ScrollChainAddress:    scrollChainAddress,
+		L1MessageQueueAddress: l1MessageQueueAddress,
+	}, l1Client)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create L1 reader: %w", err)
+	}
+
+	// 1. Sanity check user input: Make sure that the user's L1 block height is not higher than the latest finalized block number.
+	latestFinalizedL1Block, err := reader.GetLatestFinalizedBlockNumber()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get latest finalized L1 block number: %w", err)
+	}
+	if cfg.RecoveryConfig.L1BlockHeight > latestFinalizedL1Block {
+		return nil, nil, fmt.Errorf("specified L1 block height is higher than the latest finalized block number: %d > %d", cfg.RecoveryConfig.L1BlockHeight, latestFinalizedL1Block)
+	}
+
+	log.Info("Latest finalized L1 block number", "latest finalized L1 block", latestFinalizedL1Block)
+
+	// 2. Make sure that the specified batch is indeed finalized on the L1 rollup contract and is the latest finalized batch.
+	// TODO: enable check
+	//latestFinalizedBatch, err := reader.LatestFinalizedBatch(latestFinalizedL1Block)
+	//if cfg.RecoveryConfig.LatestFinalizedBatch != latestFinalizedBatch {
```

Comment:
curious about why setting `cfg.RecoveryConfig.LatestFinalizedBatch` instead of using `reader.LatestFinalizedBatch` as the start point.

Reply:
you're right, can probably remove it and make the configuration a bit easier. initially, I wanted the user to specify L1 block and the latest finalized batch so that the user knows where the (minimal) recovery process is starting from and there's no "magic" happening (e.g. if there's another batch committed in the meantime).

Reply:
I introduced another config parameter for testing purposes to override this check `force_latest_finalized_batch`. The user now still needs to specify the L1 block height and latest finalized batch to not have any surprises/magic.

Reply:
got it.